source/docs/software/commandbased/cpp-command-discussion.rst

```diff
@@ -0,0 +1,181 @@
+A Technical Discussion on C++ Commands
+======================================
+.. important:: This article serves as a technical discussion on some of the design decisions that were made when designing the new command-based framework in C++. You do not need to understand the information within this article to use the command-based framework in your robot code.
+
+.. note:: This article assumes that you have a fair understanding of advanced C++ concepts, including templates, smart pointers, inheritance, rvalue references, copy semantics, move semantics, and CRTP.
+
+This article will help you understand the reasoning behind some of the decisions made in the new command-based framework (such as the use of ``std::unique_ptr``, CRTP in the form of ``CommandHelper<Base, Derived>``, the lack of more advanced decorators that are available in Java, etc.)
+
+Ownership Model
+---------------
+The old command-based framework employed the use of raw pointers, meaning that users had to use ``new`` (resulting in manual heap allocations) in their robot code. Since there was no clear indication on who owned the commands (the scheduler, the command groups, or the user themselves), it was not apparent who was supposed to take care of freeing the memory.
+
+Several examples in the old command-based framework involved code like this:
+
+.. code-block:: c++
+
+   #include "PlaceSoda.h"
+   #include "Elevator.h"
+   #include "Wrist.h"
+
+   PlaceSoda::PlaceSoda() {
+     AddSequential(new SetElevatorSetpoint(Elevator::TABLE_HEIGHT));
+     AddSequential(new SetWristSetpoint(Wrist::PICKUP));
+     AddSequential(new OpenClaw());
+   }
+
+In the command-group above, the component commands of the command group were being heap allocated and passed into ``AddSequential`` all in the same line. This meant that that user had no reference to that object in memory and therefore had no means of freeing the allocated memory once the command group ended. The command group itself never freed the memory and neither did the command scheduler. This led to memory leaks in robot programs (i.e. memory was allocated on the heap but never freed).
+
+This glaring problem was one of the reasons for the rewrite of the framework. A comprehensive ownership model was introduced with this rewrite, along with the usage of smart pointers which will automatically free memory when they go out of scope.
+
+Commands that are not in command-groups are owned by the command scheduler whereas component commands of command groups are owned by the command group. This means that when a command or command group has finished executing, all allocated memory is automatically freed by the owner, avoiding any potential memory leaks.
```

Comment:
This isn't quite true; only default commands are owned by the scheduler.  The other commands are owned by whatever the user decides they're to be owned by.